### PR TITLE
Use Typescript and quicktype-wrapper package for code generation

### PR DESCRIPTION
**Pull request by michaelawyu**
~~To-Do: Update the GitHub repository to the official repo when [PR #70](https://github.com/googleapis/google-cloudevents/pull/70) is merged.~~

### CODE REVIEW

**Comment by grant** [comment]
Thanks for the PR @michaelawyu. Can you show a **recording** / **gif** / **image** of what it looks like when using this library?

For example, receiving this event in the http body:

```json
{
    "message": {
        "data": "SGVsbG8gdGhlcmU=",
        "messageId": "1408577928008405",
        "publishTime": "2020-08-06T22:31:14.536Z"
    },
    "subscription": "projects/MY-PROJECT/subscriptions/MY-SUB"
}
```

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,18 @@
+FROM node:12
```
Comment (grant) [inline Dockerfile:1]: I don't think a Dockerfile in the root of this repo is the right place.

I assume this is for quicktype generation. It should likely be in a separate folder like `gen`.

File: README.md [hunk 2]
```diff
@@ -1,17 +1,43 @@
...
+# The Dict may be an argument in a background Cloud Function,
+# or the payload of an HTTP request received by a Cloud Run service with event
+# triggers.
+event = MessagePublishedEvent.from_dict(event_data)
+print(event.message)
 ```
 
-Observe the new files in `src/`.
\ No newline at end of file
+## Generation
```
Comment (grant) [inline README.md:34]: Let's put this part in `.github/CONTRIBUTING.md` so we hide it from the published package.

File: gen.sh [hunk 3]
```diff
@@ -1,67 +1,32 @@
 #!/bin/bash
-# Script to generate the Protobuf classes from googleapis/google-cloudevents
 set -e
 
-echo "~ START"
-PROTOBUF_VERSION=3.12.3
+EVENTS_SPEC_REPO='https://github.com/michaelawyu/google-cloudevents'
```
Comment (grant) [inline gen.sh:4]: Use the official repo or add a TODO.

File: gen/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  if (!IN) console.error('Error in config: `IN` not set');
+  if (!OUT) console.error('Error in config: `OUT` not set');
+  if (!IN || !OUT) return;
+  if (IN.endsWith('/')) IN = IN.substring(0, IN.length - 1);
+  if (OUT.endsWith('/')) OUT = OUT.substring(0, OUT.length - 1);
+
+  const templateDirectoryPath = `${__dirname}/../../${TEMPLATE_DIRECTORY}`;
+
+  const schemasAndGenFiles = await qt.getJSONSchemasAndGenFiles(IN, LANGUAGE);
+  const allEvents: Array<[string, string, string]> = [];
```
Comment (grant) [inline gen/src/index.ts:31]: I think an array of objects would be better

Might want an interface.

```ts
interface EventData {
  package: string;
  eventName: string;
  eventDescription: string;
}
```

And then use `const allEvents: EventData[] = [];`

Objects are more common than arrays unless the data is clearly ordered.

File: gen/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+  schemasAndGenFiles.map((sg: [any, string]) => {
+    const schema = sg[0];
+    const genFile = sg[1];
```
Comment (grant) [inline gen/src/index.ts:34]: nit: We can use array destructuring to make this cleaner:

```ts
schemasAndGenFiles.map(([schema, genFile]) => {
```

**Review by grant — APPROVED** [review]
A few requested changes.

Overall looks good. I would like to see what the end experience is when using this library, we should write that down in the README.

Can you write down the features? Do you get autocompletion or docs in your IDE when accessing the request CloudEvent body parameters now?

File: Dockerfile (hunk 1 above)
Comment (michaelawyu) [reply Dockerfile]: Done.

File: gen.sh (hunk 3 above)
Comment (michaelawyu) [reply gen.sh]: The TO-DO has been listed in the PR description.
Done.

File: README.md (hunk 2 above)
Comment (michaelawyu) [reply README.md]: Done.

File: gen/src/index.ts (hunk 4 above)
Comment (michaelawyu) [reply gen/src/index.ts]: Done.

File: gen/src/index.ts (hunk 5 above)
Comment (michaelawyu) [reply gen/src/index.ts]: Done.

**Pull request merged**